src/layout/layout.rs

```diff
@@ -481,6 +475,67 @@ impl Layout {
                         element.size() | EQ(MEDIUM) | f64::from(m),
                     ])?;
                 }
+                Constraint::Length(l) => {
+                    solver.add_constraint(element.size() | EQ(STRONG) | f64::from(l))?
+                }
+                Constraint::Percentage(p) => {
+                    let percent = f64::from(p) / 100.00;
+                    solver.add_constraint(element.size() | EQ(STRONG) | (area_size * percent))?;
+                }
+                Constraint::Ratio(n, d) => {
+                    // avoid division by zero by using 1 when denominator is 0
+                    let ratio = f64::from(n) / f64::from(d.max(1));
+                    solver.add_constraint(element.size() | EQ(STRONG) | (area_size * ratio))?;
+                }
+                Constraint::Fill(_) => {
+                    // given no other constraints, fill will grow as much as possible.
+                    //
+                    // in the current implementation, this constraint will not have any effect
+                    // since in every combination of constraints, other constraints governing
+                    // element size will take a higher priority.
+                    //
+                    // this constraint is placed here only for future proofing.
+                    solver.add_constraint(element.size() | EQ(WEAK) | area_size)?;
+                }
+            }
+        }
+        // Make every `Fill` constraint proportionally equal to each other
+        // This will make `Fill` fill up empty spaces equally
+        //
+        // [Fill(1), Fill(1)]
+        // ┌──────┐┌──────┐
+        // │abcdef││abcdef│
+        // └──────┘└──────┘
+        //
+        // [Fill(1), Fill(2)]
+        // ┌──────┐┌────────────┐
+        // │abcdef││abcdefabcdef│
+        // └──────┘└────────────┘
+        //
+        // l_size / l_scale == r_size / r_scale
+        for ((&l_constraint, &l_element), (&r_constraint, &r_element)) in layout
```

Comment:
Can this be implemented by adding one extra element, which each element gets EQ to instead of constraining every combinations of elements?

e.g.
```rust
let relative_element = ...;
solver.add_constraint(
    f64::from(scaling_factor) * relative_element.size() | EQ (..) | element.size()
)
```

Reply:
I think that would work. Do you think that would be easier to understand? We are introducing an external free variable that isn't part of the layout, so at first glance it might be a little confusing.

Reply:
I think this makes the underlying concept more obvious, that there is some base value item that these scaling factors all relate to, rather than them relating to every other constraint.

Reply:
> I think this makes the underlying concept more obvious, that there is some base value item that these scaling factors all relate to, rather than them relating to every other constraint.

I think what you are suggesting is introducing a new concept, i.e. a "base value item" and then saying that the scaling factors relate to the base value item. The underlying concept as implemented right now is that the segments are "proportional" to each other.

Reply:
I tried implementing it this way. It breaks some of the tests. Specifically when there's just one `Fill(0)`.

What do we want `[Fill(0), Percentage(10)]` and `[Fill(0), Max(0)]` to do? 

When only one `Fill` is present, currently, it behaves the same regardless of the number inside it, i.e. `[Fill(0), Percentage(10)]` is the same as `[Fill(1), Percentage(10)]`, i.e. the `Percentage` constraint is respected:

```rust
let layout = Layout::default()
    .constraints([Fill(1), Percentage(20)])
    .direction(Direction::Horizontal)
    .split(Rect::new(0, 0, 100, 1));
assert_eq!(layout[0].width, 80);
assert_eq!(layout[1].width, 20);
```

If I change the implementation to what you are suggesting, when there's just one `Fill(0)`, `Fill(0)` is more strong than even a `Max` constraint, and this breaks the `Max` constraint.

```rust
let layout = Layout::default()
    .constraints([Fill(0), Max(20)])
    .direction(Direction::Horizontal)
    .split(Rect::new(0, 0, 100, 1));
assert_eq!(layout[0].width, 0);
assert_eq!(layout[1].width, 100);
```

This does not occur with `Fill(1)`:

```rust
let layout = Layout::default()
    .constraints([Fill(1), Max(20)])
    .direction(Direction::Horizontal)
    .split(Rect::new(0, 0, 100, 1));
assert_eq!(layout[0].width, 80);
assert_eq!(layout[1].width, 20);
```

When introducing a new free variable, I have to special case when there's just one `Fill(0)` to get it to do the old behavior. It feels less clean to have to do this introducing a new variable. 

I'm going to leave it as it is right now. We can revisit this and consider a refactor in the future if we find this is harder to understand.

Reply:
Does adding weak constraints EQ area.size work?
To me, based on how you've explained Fill to be needed, Fill(0) says to me "This constraint should be 0 times any other constraints that fill, it should effectively aim to be the smallest element possible."

How is Fill(0), Max(20) -> 0, 100 wrong?

Reply:
>The underlying concept as implemented right now is that the segments are "proportional" to each other.

Kinda but not - [Fill(1), Fill(2)] means proportional to a floating amount that is shared, that just happens to be the same as the first constraint. Swap it out with `Fill(2), Fill(4)` for what should have the same result, but doesn't have a unity amount to compare to. I think you can probably make the variable not free by adding [element[0] + element[1] +... | EQ(WEAK) area.size()] or something like that.

Reply:
I wrote a long comment explaining why I think it shouldn't work that way but after arguing for it, I decided to include a section on the benefits of working the way you described and I think I've convinced myself that what you suggested is what it should do. 

I can share the comment later if you'd like.

Reply:
I've changed the implementation to use an extra `base_element`.

Reply:
Here's my current thoughts on this discussion.

We are discussing two things here:

1. What should the semantics of `Proportional(0)` do?
2. What is the implementation we should use? i.e. should we use an additional variable `base_element` or not?

I'm fine with either implementation for number 2 but I have some thoughts related to mathematical numerical stability and code maintenance about number 1. So, I'd like to narrow down number 1, and then we can discuss number 2. 

There are 4 cases to consider when using `Proportional` constraints.

1. When you have code like this:

```rust
[Proportional(x), Proportional(y), ..., any_constraints_here]
```

and if the value of `x` and `y` is greater than `0`. In this case, `Proportional` works as you'd expect. It as scaled w.r.t to each other OR scaled w.r.t to a base value, the result is the same.

4. When you have code like this:

```rust
[Proportional(_), any_other_constraint_here]
```

the value inside `Proportional` shouldn't matter because there's nothing to be proportional to or relative to. Right now, i.e. with the `tuple_windows()` implementation, no constraints are added for `Proportional(_)` when there's just one of its kind.

In the `tuple_windows()` implementation, a single `Proportional(x)` behaves the same way for any value of `x`. This intuitively makes sense to me because the `x` shouldn't matter since there's nothing it can be proportional to. Or in other words, it is proportional to itself. It also behaves the same way it does when there are multiple `Proportional` constraints, i.e. they always fill up excess space. 
In the `base_element` implementation, a single `Proportional(x)` behaves differently for x equal to `0` and `x` greater than `0`. There's a special case here for `0` that I'm not particularly a fan off. Writing the code to make the semantics of this implementation behave the same as the previous requires some additional checks.

As a clarification, we can definitely get both the `tuple_windows()` implementation and the `base_element` implementation to behave in both of the above ways. We just have to choose which one we want. This is why I want to discussion the semantics first. The above is the default behavior with no additional constraints or changes. 

5. When you have code like this:

```rust
[Proportional(0), Proportional(y), ..., any_constraints_here]
```

i.e. there is just one `0` `Proportional` constraint (there is at least one other `Proportional` constraint and it and every other `Proportional` constraint is non-zero). 

There's many solutions to this problem and the solver will not be numerically stable for this formulation. Not having a stable result will cause problems with reproducibility and burden us with maintenance in the future if we want to add new constraints. I've solved this particular problem of numerical stability by forcing the `Proportional(0)` to actually be a very small value, which results in a zero sized segment in this particular case.  Even without the numerical stability fix, this would behave the same way in both `tuple_windows()` and `base_element`, i.e. it would be zero sized.

6. When you have code like this:

```rust
[Proportional(0), Proportional(0), ..., any_constraints_here]
```

and `x` and `y` are `0`, i.e. at least 2 zeros. We have to decide what we want to do here since it is technically undefined. Different solvers may return different results here. The same workaround as before works here because I've forced it to be a very small value. This makes it fill any excess space while also being proportional to any other non zero proportional constraint. Without the workaround, `tuple_windows()` behaves consistently but `base_element` does not.

`tuple_windows()`

```rust
let [a, b, c] = Rect::new(0, 0, 100, 1).split(&Layout::horizontal([
    Proportional(0),
    Fixed(1),
    Proportional(0),
]));
assert_eq!([a.width, b.width, c.width], [0, 1, 99]);

let [a, b, c] = Rect::new(0, 0, 100, 1).split(&Layout::horizontal([
    Proportional(0),
    Percentage(1),
    Proportional(0),
]));
assert_eq!([a.width, b.width, c.width], [0, 1, 99]);
```

`base_element`

```rust
let [a, b, c] = Rect::new(0, 0, 100, 1).split(&Layout::horizontal([
    Proportional(0),
    Fixed(1),
    Proportional(0),
]));
assert_eq!([a.width, b.width, c.width], [0, 1, 99]);

let [a, b, c] = Rect::new(0, 0, 100, 1).split(&Layout::horizontal([
    Proportional(0),
    Percentage(1),
    Proportional(0),
]));
assert_eq!([a.width, b.width, c.width], [0, 100, 0]);
```

`tuple_windows()` with numerical stability fix

```rust
let [a, b, c] = Rect::new(0, 0, 100, 1).split(&Layout::horizontal([
    Proportional(0),
    Fixed(1),
    Proportional(0),
]));
assert_eq!([a.width, b.width, c.width], [50, 1, 49]);

let [a, b, c] = Rect::new(0, 0, 100, 1).split(&Layout::horizontal([
    Proportional(0),
    Percentage(1),
    Proportional(0),
]));
assert_eq!([a.width, b.width, c.width], [50, 1, 49]);
```

Reply:
How about we make 0 not allowed?

I think it's generally ok to have a special case that only applies when there is one element that satisfies some condition like this, so I'm not too concerned. It's good that we call out out as we'd want to test the edge cases like this thoroughly.

I'm confused why there's multiple solutions for the prop(0), prop(1), ... case in list item 5.

Reply:
We wouldn't be able to make `0` not allowed from the user perspective, right? 

I can absolutely clamp it to `1..u16::MAX` inside the split function but if I did, these tests would change this way:

```diff
  let [a, b, c] = Rect::new(0, 0, 100, 1).split(&Layout::horizontal([
      Proportional(0),
      Proportional(1),
      Proportional(0),
  ]));
- assert_eq!([a.width, b.width, c.width], [0, 100, 0]);
+ assert_eq!([a.width, b.width, c.width], [33, 34, 33]);

  let [a, b, c, d] = Rect::new(0, 0, 100, 1).split(&Layout::horizontal([
      Proportional(0),
      Proportional(2),
      Proportional(0),
      Proportional(1),
  ]));
- assert_eq!([a.width, b.width, c.width, d.width], [0, 67, 0, 33]);
+ assert_eq!([a.width, b.width, c.width, d.width], [20, 40, 20, 20]);
```

The reason I like the behavior the way it is right now is because it is consistent with how I think about stable solutions. For example both these tests pass with the current implementation:

```rust
  let [a, b, c] = Rect::new(0, 0, 100, 1).split(&Layout::horizontal([
      Proportional(u16::MAX),
      Proportional(1),
      Percentage(20),
  ]));
  assert_eq!([a.width, b.width, c.width], [80, 0, 20]);

  let [a, b, c] = Rect::new(0, 0, 100, 1).split(&Layout::horizontal([
      Proportional(u16::MAX),
      Proportional(0),
      Percentage(20),
  ]));
  assert_eq!([a.width, b.width, c.width], [80, 0, 20]);
```